src/Lib.P2P/PeerConnection.cs

```diff
@@ -0,0 +1,433 @@
+                if (protocol != null)
+                    Protocols.Add(protocol.ToString(), protocol.ProcessMessageAsync);
+        }
+
+        /// <summary>
+        ///   Signals that the security for the connection is established.
+        /// </summary>
+        /// <remarks>
+        ///   This can be awaited.
+        /// </remarks>
+        public TaskCompletionSource<bool> SecurityEstablished { get; } = new TaskCompletionSource<bool>();
+
+        /// <summary>
+        ///   Signals that the muxer for the connection is established.
+        /// </summary>
+        /// <remarks>
+        ///   This can be awaited.
+        /// </remarks>
+        public TaskCompletionSource<Muxer> MuxerEstablished { get; } = new TaskCompletionSource<Muxer>();
+
+        /// <summary>
+        ///   Signals that the identity of the remote endpoint is established.
+        /// </summary>
+        /// <remarks>
+        ///   This can be awaited.
+        /// </remarks>
+        /// <remarks>
+        ///   The data in <see cref="RemotePeer"/> is not complete until
+        ///   the identity is establish.
+        /// </remarks>
+        public TaskCompletionSource<Peer> IdentityEstablished { get; } = new TaskCompletionSource<Peer>();
+
+        /// <summary>
+        ///   When the connection was last used.
+        /// </summary>
+        public DateTime LastUsed => _statsStream.LastUsed;
+
+        /// <summary>
+        ///   Number of bytes read over the connection.
+        /// </summary>
+        public long BytesRead => _statsStream.BytesRead;
+
+        /// <summary>
+        ///   Number of bytes written over the connection.
+        /// </summary>
+        public long BytesWritten => _statsStream.BytesWritten;
+
+        /// <summary>
+        ///  Establish the connection with the remote node.
+        /// </summary>
+        /// <param name="securityProtocols"></param>
+        /// <param name="cancel"></param>
+        /// <remarks>
+        ///   This should be called when the local peer wants a connection with
+        ///   the remote peer.
+        /// </remarks>
+        public async Task InitiateAsync(IEnumerable<IEncryptionProtocol> securityProtocols,
+            CancellationToken cancel = default)
+        {
+            await EstablishProtocolAsync("/multistream/", cancel).ConfigureAwait(false);
+
+            // Find the first security protocol that is also supported by the remote.
+            var exceptions = new List<Exception>();
+            foreach (var protocol in securityProtocols)
+            {
+                try
+                {
+                    await EstablishProtocolAsync(protocol.ToString(), cancel).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    exceptions.Add(e);
+                    continue;
+                }
+
+                await protocol.EncryptAsync(this, cancel).ConfigureAwait(false);
+                break;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor the code in order to remove this break statement.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)